File: Treatment Start Patterns and Adherence Modeling for Clinical Interventions/scripts/python/comprehensive_eda.py
"""
Comprehensive Exploratory Data Analysis Script

This script performs thorough exploratory data analysis of the treatment starts dataset.
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import warnings
warnings.filterwarnings('ignore')

def load_and_preprocess_data(file_path):
    """Load and preprocess the dataset."""
    df = pd.read_csv(file_path)
    df['TreatmentStart'] = pd.to_datetime(df['TreatmentStart'], format='%m/%d/%y')
    df['Year'] = df['TreatmentStart'].dt.year
    df['Month'] = df['TreatmentStart'].dt.month
    df['Day'] = df['TreatmentStart'].dt.day
    df['MonthName'] = df['TreatmentStart'].dt.strftime('%B')
    df['Weekday'] = df['TreatmentStart'].dt.day_name()
    df['Quarter'] = df['TreatmentStart'].dt.quarter
    return df

def data_quality_assessment(df):
    """Assess data quality."""
    print("=" * 60)
    print("DATA QUALITY ASSESSMENT")
    print("=" * 60)
    
    print(f"\nDataset shape: {df.shape}")
    print(f"Missing values:\n{df.isnull().sum()}")
    print(f"\nData types:\n{df.dtypes}")
    print(f"\nDate range: {df['TreatmentStart'].min()} to {df['TreatmentStart'].max()}")
    
    # Outlier detection
    Q1 = df['Dosage'].quantile(0.25)
    Q3 = df['Dosage'].quantile(0.75)
    IQR = Q3 - Q1
    outliers = df[(df['Dosage'] < Q1 - 1.5*IQR) | (df['Dosage'] > Q3 + 1.5*IQR)]
    print(f"\nOutliers detected: {len(outliers)}")
    if len(outliers) > 0:
        print("Outlier details:")
        print(outliers[['PatientID', 'Drug', 'Dosage']])
    
    return outliers

def descriptive_statistics(df):
    """Compute descriptive statistics."""
    print("\n" + "=" * 60)
    print("DESCRIPTIVE STATISTICS")
    print("=" * 60)
    
    print("\nNumerical variables:")
    print(df['Dosage'].describe())
    print(f"\nSkewness: {df['Dosage'].skew():.4f}")
    print(f"Kurtosis: {df['Dosage'].kurtosis():.4f}")
    
    print("\nCategorical variables:")
    print("Drug distribution:")
    print(df['Drug'].value_counts())
    print(f"\nProportions:\n{df['Drug'].value_counts(normalize=True) * 100}")

def temporal_analysis(df, output_dir='outputs'):
    """Analyze temporal patterns."""
    print("\n" + "=" * 60)
    print("TEMPORAL ANALYSIS")
    print("=" * 60)
    
    month_order = ['January', 'February', 'March', 'April', 'May', 'June']
    monthly_counts = df['MonthName'].value_counts().reindex(month_order, fill_value=0)
    
    print("\nTreatment starts by month:")
    print(monthly_counts)
    
    print("\nMean dosage by month:")
    mean_dosage_month = df.groupby('MonthName')['Dosage'].mean().reindex(month_order, fill_value=0)
    print(mean_dosage_month)
    
    # Visualizations
    fig, axes = plt.subplots(2, 2, figsize=(15, 10))
    
    # Monthly counts
    axes[0, 0].bar(range(len(monthly_counts)), monthly_counts.values)
    axes[0, 0].set_xticks(range(len(monthly_counts)))
    axes[0, 0].set_xticklabels(monthly_counts.index, rotation=45)
    axes[0, 0].set_title('Treatment Starts by Month')
    axes[0, 0].set_ylabel('Count')
    
    # Mean dosage by month
    axes[0, 1].plot(range(len(mean_dosage_month)), mean_dosage_month.values, marker='o')
    axes[0, 1].set_xticks(range(len(mean_dosage_month)))
    axes[0, 1].set_xticklabels(mean_dosage_month.index, rotation=45)
    axes[0, 1].set_title('Mean Dosage by Month')
    axes[0, 1].set_ylabel('Mean Dosage')
    
    # Timeline
    df_sorted = df.sort_values('TreatmentStart')
    axes[1, 0].plot(df_sorted['TreatmentStart'], range(len(df_sorted)), marker='o', markersize=4)
    axes[1, 0].set_title('Treatment Starts Timeline')
    axes[1, 0].set_xlabel('Date')
    axes[1, 0].set_ylabel('Cumulative Count')
    axes[1, 0].tick_params(axis='x', rotation=45)
    
    # Drug usage by month
    drug_month = pd.crosstab(df['MonthName'], df['Drug']).reindex(month_order, fill_value=0)
    sns.heatmap(drug_month, annot=True, fmt='d', cmap='YlOrRd', ax=axes[1, 1])
    axes[1, 1].set_title('Drug Usage by Month')
    
    plt.tight_layout()
    plt.savefig(f'{output_dir}/temporal_analysis.png', dpi=300, bbox_inches='tight')
    plt.close()

def relationship_analysis(df, output_dir='outputs'):
    """Analyze relationships between variables."""
    print("\n" + "=" * 60)
    print("RELATIONSHIP ANALYSIS")
    print("=" * 60)
    
    # Correlation
    df_encoded = df.copy()
    df_encoded['Drug_encoded'] = df_encoded['Drug'].map({'Cisplatin': 0, 'Nivolumab': 1})
    corr_matrix = df_encoded[['Dosage', 'Drug_encoded', 'Month']].corr()
    
    print("\nCorrelation matrix:")
    print(corr_matrix)
    
    # Visualizations
    fig, axes = plt.subplots(2, 2, figsize=(15, 10))
    
    # Dosage by Drug
    sns.boxplot(data=df, x='Drug', y='Dosage', ax=axes[0, 0])
    axes[0, 0].set_title('Dosage Distribution by Drug')
    
    # Correlation heatmap
    sns.heatmap(corr_matrix, annot=True, fmt='.3f', cmap='coolwarm', center=0, ax=axes[0, 1])
    axes[0, 1].set_title('Correlation Matrix')
    
    # Scatter plot
    scatter = axes[1, 0].scatter(df['Month'], df['Dosage'], 
                                c=df_encoded['Drug_encoded'], cmap='viridis', alpha=0.6)
    axes[1, 0].set_title('Month vs Dosage (colored by Drug)')
    axes[1, 0].set_xlabel('Month')
    axes[1, 0].set_ylabel('Dosage')
    plt.colorbar(scatter, ax=axes[1, 0])
    
    # Distribution comparison
    for drug in df['Drug'].unique():
        drug_data = df[df['Drug'] == drug]
        axes[1, 1].hist(drug_data['Dosage'], alpha=0.6, label=drug, bins=10)
    axes[1, 1].set_title('Dosage Distribution by Drug')
    axes[1, 1].set_xlabel('Dosage')
    axes[1, 1].set_ylabel('Frequency')
    axes[1, 1].legend()
    
    plt.tight_layout()
    plt.savefig(f'{output_dir}/relationship_analysis.png', dpi=300, bbox_inches='tight')
    plt.close()

def patient_analysis(df):
    """Analyze patient-level patterns."""
    print("\n" + "=" * 60)
    print("PATIENT ANALYSIS")
    print("=" * 60)
    
    patient_stats = df.groupby('PatientID').agg({
        'Drug': 'count',
        'Dosage': ['mean', 'sum', 'min', 'max']
    }).round(2)
    
    patient_stats.columns = ['Treatment_Count', 'Mean_Dosage', 'Total_Dosage', 
                            'Min_Dosage', 'Max_Dosage']
    
    print(f"\nNumber of unique patients: {df['PatientID'].nunique()}")
    print(f"Patients with multiple treatments: {len(patient_stats[patient_stats['Treatment_Count'] > 1])}")
    print("\nPatient Statistics (top 10):")
    print(patient_stats.sort_values('Treatment_Count', ascending=False).head(10))

def generate_summary(df, outliers):
    """Generate EDA summary."""
    print("\n" + "=" * 60)
    print("EDA SUMMARY")
    print("=" * 60)
    
    print(f"\n1. Dataset Overview:")
    print(f"   - Total records: {len(df)}")
    print(f"   - Unique patients: {df['PatientID'].nunique()}")
    print(f"   - Date range: {df['TreatmentStart'].min().date()} to {df['TreatmentStart'].max().date()}")
    
    print(f"\n2. Drug Distribution:")
    for drug, count in df['Drug'].value_counts().items():
        print(f"   - {drug}: {count} treatments ({count/len(df)*100:.1f}%)")
    
    print(f"\n3. Dosage Statistics:")
    print(f"   - Mean: {df['Dosage'].mean():.2f}")
    print(f"   - Median: {df['Dosage'].median():.2f}")
    print(f"   - Std Dev: {df['Dosage'].std():.2f}")
    print(f"   - Range: {df['Dosage'].min()} - {df['Dosage'].max()}")
    
    print(f"\n4. Data Quality:")
    print(f"   - Missing values: {df.isnull().sum().sum()}")
    print(f"   - Outliers detected: {len(outliers)}")
    
    print(f"\n5. Key Findings:")
    print("   - Dataset contains treatment starts for two drugs: Cisplatin and Nivolumab")
    print("   - Some patients have multiple treatment records")
    print("   - Dosage varies significantly between drugs")
    print("   - Treatment patterns show temporal variations")

def main():
    """Main function."""
    import os
    
    # Create output directory
    output_dir = 'outputs'
    os.makedirs(output_dir, exist_ok=True)
    
    # Load data
    file_path = '../../data/mock_treatment_starts_2016.csv'
    df = load_and_preprocess_data(file_path)
    
    # Perform analyses
    outliers = data_quality_assessment(df)
    descriptive_statistics(df)
    temporal_analysis(df, output_dir)
    relationship_analysis(df, output_dir)
    patient_analysis(df)
    generate_summary(df, outliers)
    
    print("\n" + "=" * 60)
    print("EDA COMPLETE")
    print("=" * 60)

if __name__ == "__main__":
    main()

File: Treatment Start Patterns and Adherence Modeling for Clinical Interventions/scripts/r/comprehensive_eda.R
# Comprehensive Exploratory Data Analysis Script
# This script performs thorough exploratory data analysis of the treatment starts dataset.

# Load libraries
library(dplyr)
library(ggplot2)
library(tidyr)
library(lubridate)
library(gridExtra)
library(moments)
library(VIM)

# Helper function
`%+%` <- function(a, b) paste0(a, b)

# Load and preprocess data
load_and_preprocess_data <- function(file_path) {
  df <- read.csv(file_path, stringsAsFactors = FALSE)
  df$TreatmentStart <- as.Date(df$TreatmentStart, format = "%m/%d/%y")
  df$Year <- year(df$TreatmentStart)
  df$Month <- month(df$TreatmentStart)
  df$Day <- day(df$TreatmentStart)
  df$MonthName <- month.name[df$Month]
  df$Weekday <- weekdays(df$TreatmentStart)
  df$Quarter <- quarter(df$TreatmentStart)
  return(df)
}

# Data quality assessment
data_quality_assessment <- function(df) {
  cat("=" %+% rep("=", 59), "\n")
  cat("DATA QUALITY ASSESSMENT\n")
  cat("=" %+% rep("=", 59), "\n")
  
  cat("\nDataset shape:", dim(df), "\n")
  cat("Missing values:\n")
  print(colSums(is.na(df)))
  cat("Date range:", min(df$TreatmentStart), "to", max(df$TreatmentStart), "\n")
  
  # Outlier detection
  Q1 <- quantile(df$Dosage, 0.25)
  Q3 <- quantile(df$Dosage, 0.75)
  IQR_val <- IQR(df$Dosage)
  outliers <- df[df$Dosage < (Q1 - 1.5*IQR_val) | df$Dosage > (Q3 + 1.5*IQR_val), ]
  cat("\nOutliers detected:", nrow(outliers), "\n")
  if (nrow(outliers) > 0) {
    cat("Outlier details:\n")
    print(outliers[, c("PatientID", "Drug", "Dosage")])
  }
  
  return(outliers)
}

# Descriptive statistics
descriptive_statistics <- function(df) {
  cat("\n" %+% "=" %+% rep("=", 59), "\n")
  cat("DESCRIPTIVE STATISTICS\n")
  cat("=" %+% rep("=", 59), "\n")
  
  cat("\nNumerical variables:\n")
  print(summary(df$Dosage))
  cat("Skewness:", skewness(df$Dosage), "\n")
  cat("Kurtosis:", kurtosis(df$Dosage), "\n")
  
  cat("\nCategorical variables:\n")
  cat("Drug distribution:\n")
  print(table(df$Drug))
  cat("Proportions:\n")
  print(prop.table(table(df$Drug)) * 100)
}

# Temporal analysis
temporal_analysis <- function(df, output_dir = "outputs") {
  cat("\n" %+% "=" %+% rep("=", 59), "\n")
  cat("TEMPORAL ANALYSIS\n")
  cat("=" %+% rep("=", 59), "\n")
  
  month_order <- month.name[1:6]
  monthly_counts <- df %>%
    group_by(MonthName) %>%
    summarise(Count = n()) %>%
    arrange(match(MonthName, month_order))
  
  cat("\nTreatment starts by month:\n")
  print(monthly_counts)
  
  cat("\nMean dosage by month:\n")
  mean_dosage_month <- df %>%
    group_by(MonthName) %>%
    summarise(Mean_Dosage = mean(Dosage)) %>%
    arrange(match(MonthName, month_order))
  print(mean_dosage_month)
  
  # Visualizations
  png(paste0(output_dir, "/temporal_analysis.png"), width = 1200, height = 800)
  
  p1 <- ggplot(monthly_counts, aes(x = MonthName, y = Count)) +
    geom_bar(stat = "identity", fill = "steelblue") +
    labs(title = "Treatment Starts by Month", x = "Month", y = "Count") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  p2 <- ggplot(mean_dosage_month, aes(x = MonthName, y = Mean_Dosage)) +
    geom_line(group = 1, color = "teal", size = 1) +
    geom_point(color = "teal", size = 3) +
    labs(title = "Mean Dosage by Month", x = "Month", y = "Mean Dosage") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  df_sorted <- df[order(df$TreatmentStart), ]
  df_sorted$Cumulative <- 1:nrow(df_sorted)
  p3 <- ggplot(df_sorted, aes(x = TreatmentStart, y = Cumulative)) +
    geom_line() +
    geom_point(size = 2) +
    labs(title = "Treatment Starts Timeline", x = "Date", y = "Cumulative Count") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  drug_month <- table(df$MonthName, df$Drug)
  p4 <- ggplot(as.data.frame(drug_month), aes(x = Var1, y = Var2, fill = Freq)) +
    geom_tile() +
    scale_fill_gradient(low = "white", high = "red") +
    labs(title = "Drug Usage by Month", x = "Month", y = "Drug", fill = "Count") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  grid.arrange(p1, p2, p3, p4, nrow = 2, ncol = 2)
  dev.off()
}

# Relationship analysis
relationship_analysis <- function(df, output_dir = "outputs") {
  cat("\n" %+% "=" %+% rep("=", 59), "\n")
  cat("RELATIONSHIP ANALYSIS\n")
  cat("=" %+% rep("=", 59), "\n")
  
  # Correlation
  df_encoded <- df
  df_encoded$Drug_encoded <- ifelse(df_encoded$Drug == "Cisplatin", 0, 1)
  corr_matrix <- cor(df_encoded[, c("Dosage", "Drug_encoded", "Month")])
  
  cat("\nCorrelation matrix:\n")
  print(corr_matrix)
  
  # Visualizations
  png(paste0(output_dir, "/relationship_analysis.png"), width = 1200, height = 800)
  
  p1 <- ggplot(df, aes(x = Drug, y = Dosage)) +
    geom_boxplot() +
    labs(title = "Dosage Distribution by Drug") + theme_minimal()
  
  library(corrplot)
  corrplot(corr_matrix, method = "color", type = "upper",
           addCoef.col = "black", tl.col = "black", tl.srt = 45)
  
  p3 <- ggplot(df_encoded, aes(x = Month, y = Dosage, color = factor(Drug_encoded))) +
    geom_point(alpha = 0.6, size = 3) +
    scale_color_manual(values = c("0" = "blue", "1" = "red"),
                       labels = c("0" = "Cisplatin", "1" = "Nivolumab")) +
    labs(title = "Month vs Dosage (colored by Drug)",
         x = "Month", y = "Dosage", color = "Drug") + theme_minimal()
  
  p4 <- ggplot(df, aes(x = Dosage, fill = Drug)) +
    geom_histogram(alpha = 0.6, bins = 10, position = "identity") +
    labs(title = "Dosage Distribution by Drug", x = "Dosage", y = "Frequency") +
    theme_minimal()
  
  grid.arrange(p1, p3, p4, nrow = 2, ncol = 2)
  dev.off()
}

# Patient analysis
patient_analysis <- function(df) {
  cat("\n" %+% "=" %+% rep("=", 59), "\n")
  cat("PATIENT ANALYSIS\n")
  cat("=" %+% rep("=", 59), "\n")
  
  patient_stats <- df %>%
    group_by(PatientID) %>%
    summarise(
      Treatment_Count = n(),
      Mean_Dosage = mean(Dosage),
      Total_Dosage = sum(Dosage),
      Min_Dosage = min(Dosage),
      Max_Dosage = max(Dosage)
    )
  
  cat("\nNumber of unique patients:", length(unique(df$PatientID)), "\n")
  cat("Patients with multiple treatments:", 
      sum(patient_stats$Treatment_Count > 1), "\n")
  cat("\nPatient Statistics (top 10):\n")
  print(head(patient_stats[order(patient_stats$Treatment_Count, decreasing = TRUE), ], 10))
}

# Generate summary
generate_summary <- function(df, outliers) {
  cat("\n" %+% "=" %+% rep("=", 59), "\n")
  cat("EDA SUMMARY\n")
  cat("=" %+% rep("=", 59), "\n")
  
  cat("\n1. Dataset Overview:\n")
  cat("   - Total records:", nrow(df), "\n")
  cat("   - Unique patients:", length(unique(df$PatientID)), "\n")
  cat("   - Date range:", min(df$TreatmentStart), "to", max(df$TreatmentStart), "\n")
  
  cat("\n2. Drug Distribution:\n")
  drug_counts <- table(df$Drug)
  for (i in 1:length(drug_counts)) {
    cat("   -", names(drug_counts)[i], ":", drug_counts[i],
        "treatments (", round(drug_counts[i]/nrow(df)*100, 1), "%)\n")
  }
  
  cat("\n3. Dosage Statistics:\n")
  cat("   - Mean:", mean(df$Dosage), "\n")
  cat("   - Median:", median(df$Dosage), "\n")
  cat("   - Std Dev:", sd(df$Dosage), "\n")
  cat("   - Range:", min(df$Dosage), "-", max(df$Dosage), "\n")
  
  cat("\n4. Data Quality:\n")
  cat("   - Missing values:", sum(is.na(df)), "\n")
  cat("   - Outliers detected:", nrow(outliers), "\n")
  
  cat("\n5. Key Findings:\n")
  cat("   - Dataset contains treatment starts for two drugs: Cisplatin and Nivolumab\n")
  cat("   - Some patients have multiple treatment records\n")
  cat("   - Dosage varies significantly between drugs\n")
  cat("   - Treatment patterns show temporal variations\n")
}

# Main function
main <- function() {
  # Create output directory
  if (!dir.exists("outputs")) {
    dir.create("outputs")
  }
  
  # Load data
  file_path <- "../../data/mock_treatment_starts_2016.csv"
  df <- load_and_preprocess_data(file_path)
  
  # Perform analyses
  outliers <- data_quality_assessment(df)
  descriptive_statistics(df)
  temporal_analysis(df)
  relationship_analysis(df)
  patient_analysis(df)
  generate_summary(df, outliers)
  
  cat("\n" %+% "=" %+% rep("=", 59), "\n")
  cat("EDA COMPLETE\n")
  cat("=" %+% rep("=", 59), "\n")
}

# Run main function
if (!interactive()) {
  main()
}


